authentication/KratosApiAuthentication.js

```diff
@@ -0,0 +1,135 @@
+/*
+ * This script is used to authenticate a user using Ory Kratos self-service API for clients without browsers.
+ *
+ * Authentication Verification can be configured with the following rule:
+ * - Verification Strategy: Poll the Specified URL
+ * - Regex Pattern used to identify Logged In message: \Qactive.*
+ * - URL to poll: <Kratos Base URL>/sessions/whoami
+ *
+ * Zap must be configured to use HTTP header-based session management with the value: {%json:session_token%}
+ *
+ * @author Edouard Maleix <ed@getlarge.eu>
+ * @see https://www.ory.sh/docs/kratos/self-service/flows/user-login#login-for-api-clients-and-clients-without-browsers
+ */
+
+const HttpRequestHeader = Java.type(
+  "org.parosproxy.paros.network.HttpRequestHeader"
+);
+const HttpHeader = Java.type("org.parosproxy.paros.network.HttpHeader");
+const URI = Java.type("org.apache.commons.httpclient.URI");
+const AuthenticationHelper = Java.type(
+  "org.zaproxy.zap.authentication.AuthenticationHelper"
+);
+
+/**
+ * @typedef {Object} AuthHelper
+ * @property {function(): Object} prepareMessage - Prepares an HTTP message.
+ * @property {function(Object, boolean=): void} sendAndReceive - Sends the HTTP message and receives the response.
+ */
+
+/**
+ * @typedef {Object} ParamsValues
+ * @property {function(string): string} get - Gets the value of a parameter.
+ * @property {function(string): void} set - Sets the value of a parameter.
+ */
+
+/**
+ * @typedef {Object} Credentials
+ * @property {function(string): string} getParam - Gets the value of a parameter. The param	names are the ones returned by the getCredentialsParamsNames() below
+ */
+
+/**
+ * @param {AuthHelper} helper - The authentication helper object provided by ZAP.
+ * @param {ParamsValues} paramsValues - The map of parameter values configured in the Session Properties - Authentication panel.
+ * @param {Credentials} credentials - an object containing the credentials values, as configured in the Session Properties - Users panel.
+ * @returns {Object} The HTTP message used to perform the authentication.
+ */
+function authenticate(helper, paramsValues, credentials) {
+  print("Authenticating via Ory Kratos...");
+
+  // Step 1: Initialize the login flow
+  const kratosBaseUri = paramsValues.get("Kratos Base URL");
+  const initLoginUri = new URI(
+    kratosBaseUri + "/self-service/login/api",
+    false
```

Comment:
Would be better if this was true, having the URI already escaped/encoded avoids surprises.

Reply:
Honestly here I just took inspiration from the other Authentication snippets which all seems to set false as a value for the second parameter.
If I understand correctly, you would suggest to do something like:
```js
new URI(encodeURI(kratosBaseUri + "/self-service/login/api"), true);
```
for all URI instantiation?

Reply:
Yes.